pkg/bpf/map_linux.go

```diff
@@ -88,6 +94,94 @@ type cacheEntry struct {
 	LastError     error
 }
 
+type historyEntry struct {
+	Timestamp time.Time
+	Key       MapKey
+	AllKeys   bool
+	Value     MapValue
+	Action    DesiredAction
+	Err       error
+}
+
+// HistoryManager is an utility which stores history of operations on the BPF
+// map.
+type HistoryManager struct {
+	i       int
+	entries []*historyEntry
+	mutex   lock.RWMutex
+}
+
+func newHistoryManager() *HistoryManager {
+	entries := make([]*historyEntry, 0, maxHistoryEntries)
+	return &HistoryManager{
+		entries: entries,
+		mutex:   lock.RWMutex{},
+	}
+}
+
+func (hm *HistoryManager) addEntry(key MapKey, value MapValue,
+	action DesiredAction, err error) {
+	entry := &historyEntry{
+		Timestamp: time.Now(),
+		Key:       key,
+		Value:     value,
+		Action:    action,
+		Err:       err,
+	}
+
+	hm.mutex.Lock()
+	defer hm.mutex.Unlock()
+
+	// It max amount of entries is exceeded, get the subslice without the
+	// first element and then append the new entry.
+	if hm.i >= maxHistoryEntries {
+		hm.entries = hm.entries[1:maxHistoryEntries]
+		hm.entries = append(hm.entries, entry)
+		return
+	}
+	hm.i++
+	hm.entries = append(hm.entries, entry)
+}
+
+func (hm *HistoryManager) addDeleteAllEntry(err error) {
+	entry := &historyEntry{
+		Timestamp: time.Now(),
+		AllKeys:   true,
+		Action:    Delete,
+		Err:       err,
+	}
+
+	hm.mutex.Lock()
+	defer hm.mutex.Unlock()
+	hm.entries = append(hm.entries, entry)
```

Comment:
We probably want to do the same bookkeeping here to avoid exceeding max entries?

Reply:
Thanks for pointing that out. :man_facepalming: 

I moved the whole bookkeeping logic to ´addEntryMeta` and all the other methods are calling it.